src/services/addressProvider.ts

```diff
@@ -40,15 +48,44 @@ export class AddressProvider<T extends ChainId> extends Service {
       case 42161:
         return "0xcAd10033C86B0C1ED6bfcCAa2FF6779938558E9f";
       default:
-        throw new Error("Unsupported chain id");
+        throw new Error(`Unsupported chain id: ${chainId}`);
     }
   }
 
-  async addressById(id: ContractAddressId): Promise<string> {
+  async addressById(id: ContractAddressId): Promise<Address> {
+    const { address, timestamp } = this.cachedAddressesById.get(id) || {};
+
+    if (address && this.isCacheFresh({ timestamp })) {
+      return address;
+    }
+
     return this.contract.read.addressById(id);
   }
 
+  setCachedAddressById(id: ContractAddressId, address: Address) {
+    const NOW = new Date(Date.now()).getTime();
+
+    if (!this.cachedAddressesById.has(id)) {
+      this.cachedAddressesById.set(id, { address, timestamp: NOW });
+      return;
+    }
+
+    const { timestamp } = this.cachedAddressesById.get(id) || {};
+
+    if (!this.isCacheFresh({ timestamp })) {
+      this.cachedAddressesById.set(id, { address, timestamp: NOW });
+    }
+  }
+
   async addressesMetadataByIdStartsWith(prefix: string): Promise<AddressMetadata[]> {
     return this.contract.read.addressById(prefix).then(structArray);
```

Comment:
Wondering if this function is finalised? `addressById` returns a string, then `structArray` tries to map through it, added a test case that expect this function to throw https://github.com/yearn/yearn-sdk/pull/279/files#diff-cb54b208bf2147ced294ff156e88f046bb4301885610276eee7017763c9084a1R124

Reply:
It's not used atm I believe, it was just added because it's a function that's present on the contract, so the implementation might be incorrect

Reply:
The function is the same though, they are both `addressById`

Reply:
you're right, it's not calling the correct function, should be using `this.contract.read. addressesMetadataByIdStartsWith`

Reply:
Worth to open an issue for this so we don't forget. Will resolve this for now